Source/DafnyCore/Compilers/GoLang/Compiler-go.cs

```diff
@@ -262,23 +270,39 @@ private struct Import {
       var staticFieldInitWriter = w.NewNamedBlock("var {0} = {1}", FormatCompanionName(name), FormatCompanionTypeName(name));
 
       if (includeEquals) {
-        // This Equals() is so simple that we could just use == instead, but uniformity is good and it'll get inlined anyway.
 
         w.WriteLine();
-        var wEquals = w.NewNamedBlock("func (_this *{0}) Equals(other *{0}) bool", name);
-        wEquals.WriteLine("return _this == other");
+        // TODO-RS: HACK
+        if (name.EndsWith("Sequence")) {
+          var wEquals = w.NewNamedBlock("func (_this *{0}) Equals(other Sequence) bool", name);
+          wEquals.WriteLine($"return {DafnySequenceCompanion}.Equal(_this, other)");
```

Comment:
It's not Go-specific, it's necessary to allow the `Sequence<T>` trait and implementing classes to implement equality that isn't the normal reference equality. The hack is currently to special case this exact type (and it needs to be more precisely looking for `dafny.Sequence<T>` so we can rely on the fact that [users aren't supposed to define their own top-level `dafny` module](https://github.com/dafny-lang/dafny/blob/master/docs/StyleGuide/Style-Guide.md#module-names)).

A cleaner alternative would be to have an attribute like `{:customEquals}` we could tag the trait/classes with, but I only want to do that if we can prevent user code from using it somehow, since it would only be half-baked and unsafe in general.

Reply:
I suspect it's fine to leave it as-is for now, but in the long term maybe it makes sense to add a `SupportsCustomEquality` property on `Type` to use in all of the compilers. It might essentially be the same logic at first, but located somewhere more easily shared by multiple compiler backends.